Automation API User Guide (#65)
pulumi/pulumi-hugo: themes/default/content/docs/guides/automation-api/concepts-terminology.md

---
title: Core Concepts and Terminology
meta_desc: This page contains an overview of core concepts when using Automation API.
linktitle: Automation API Concepts
weight: 1

menu:
  userguides:
    parent: automation-api
---

Automation API lets you define a Pulumi program as a function within your codebase rather than in a separate project and use methods to get and set configuration parameters programmatically.

The package can be used for a number of use cases:

- Driving pulumi deployments within CI/CD workflows
- Integration testing
- Multi-stage deployments such as blue-green deployment patterns
- Deployments involving application code like database migrations

> - Building higher level tools, custom CLIs over pulumi, etc

Reviewer: ```suggestion
- Building higher level tools, custom CLIs over Pulumi, etc.
```

> - Using pulumi behind a REST or GRPC API

Reviewer: ```suggestion
- Using Pulumi behind a REST or GRPC API
```

- Debugging Pulumi programs (by using a single main entrypoint with "inline" programs)

## Workspace

To enable a broad range of runtime customization the API defines a Workspace interface. A Workspace is the execution context containing a single Pulumi project, a program, and multiple stacks. Workspaces are used to manage the execution environment, providing various utilities such as plugin installation, environment configuration (`$PULUMI_HOME`), and creation, deletion, and listing of Stacks.

> Custom Workspace implementations can be used to store Project and Stack settings as well as Config in a different format, such as an in-memory data structure, a shared persistent SQL database, or cloud object storage. Regardless of the backing Workspace implementation, the Pulumi SaaS Console will still be able to display configuration applied to updates as it does with the local version of the Workspace today.

Reviewer: ```suggestion
Custom workspace implementations can be used to store project, stack and configuration values in a different format, such as an in-memory data structure, a shared persistent SQL database, or cloud object storage. Regardless of the backing workspace implementation, the Pulumi Console will still be able to display configuration applied to updates as it does with the local version of the workspace today.
```

Can you add a bit more clarity on "...still be able to display configuration applied to updates"? Do you mean configuration changes applied to updates?

### LocalWorkspace

> A default implementation of workspace is provided as LocalWorkspace. This implementation relies on `Pulumi.yaml` and `Pulumi.[stack].yaml` as the intermediate format for Project and Stack settings. Modifying `ProjectSettings` will alter the Workspace `Pulumi.yaml` file, and setting config on a Stack will modify the `Pulumi.[stack].yaml` file. This is identical to the behavior of Pulumi CLI driven workspaces.

Reviewer: Not sure if I changed the meaning of the first sentence, so let me know if it's not quite right.

```suggestion
The default implementation of a workspace is known as a LocalWorkspace. This implementation relies on `Pulumi.yaml` and `Pulumi.[stack].yaml` as the intermediate format for Project and Stack settings. Modifying `ProjectSettings` will alter the workspace `Pulumi.yaml` file, and setting config on a stack will modify the `Pulumi.[stack].yaml` file. This is identical to the behavior of Pulumi CLI-driven workspaces.
```

## Stack

> Stack is an isolated, independently configurable instance of a Pulumi program. Stack exposes methods for the full pulumi lifecycle (up/preview/refresh/destroy), as well as managing configuration. Multiple Stacks are commonly used to denote different phases of development (such as development, staging and production) or feature branches (such as feature-x-dev, jane-feature-x-dev).

Reviewer: I suggest just nixing this section and/or just linking to /docs/intro/concepts/stack/.

Author: Yeah, good point. I guess in this specific instance we are talking about the class `Stack` rather than the concept of a "stack". This is part of the trouble I was having writing this section, because a lot of these concepts are familiar, the distinction is just that we're talking specifically about software `Class`es rather than the more general concepts. But maybe that distinction isn't necessary.

> ### Local Program

Reviewer: ```suggestion
## Local Program
```

> A Local Program is a traditional Pulumi CLI-driven program with its own directory, a `Pulumi.yaml` file, along with a file that defines the Pulumi program. Automation API can be used to drive these programs, as well as the CLI.

Reviewer: ```suggestion
A local program is a traditional Pulumi CLI-driven program with its own directory, a `Pulumi.yaml` file, along with a file that defines the Pulumi program. Automation API can be used to drive these programs, as well as the CLI.
```

> ### Inline Program

Reviewer: ```suggestion
## Inline Program
```

> Unlike traditional Pulumi programs, inline programs don't require a separate package on disk, with its own file and a `Pulumi.yaml`. Inline programs are just functions that can be authored in the same file as your Automation API program or be imported from anther package.

Reviewer: ```suggestion
Unlike traditional Pulumi programs, inline programs don't require a separate package on disk, with its own file and a `Pulumi.yaml`. Inline programs are simply functions that can be authored in the same file as your Automation API program or be imported from anther package.
```
